### PR TITLE
Thrift timeout

### PR DESCRIPTION
**Pull request by sakshia125**
Add timeouts and retries to thrift clients

### CODE REVIEW

**Comment by stagraqubole** [comment]
difficult to review with this based over non-local1 and review opened against master.

Is it possible to re-open the PR against non-local1?

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                bookKeeperClient.transport.open();
+                return callable.call();
+            }
+            catch (TTransportException e1) {
+                LOG.info("Error while reconnecting");
+                errors++;
+            }
+        }
+
+        if (errors >= maxRetries) {
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: this if is not needed, this will always be true when we come here

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        super(new TBinaryProtocol(client.transport));
+        this.bookKeeperClient = client;
+        this.maxRetries = maxRetries;
+    }
+
+    @Override
+    public List<Location> getCacheStatus(final String remotePath, final long fileLength, final long lastModified, final long startBlock, final long endBlock, final int clusterType)
+    {
+        try {
+            return bookKeeperClient.getCacheStatus(remotePath, fileLength, lastModified, startBlock, endBlock, clusterType);
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: why this call? just do retryConnection call

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void setAllCached(final String remotePath, final long fileLength, final long lastModified, final long startBlock, final long endBlock)
+    {
+        try {
+            bookKeeperClient.setAllCached(remotePath, fileLength, lastModified, startBlock, endBlock);
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: same like above, no need of this call

File: rubix-core/src/main/java/com/qubole/rubix/core/CachingFileSystem.java [hunk 4]
```diff
@@ -115,9 +118,10 @@ public FSDataInputStream open(Path path, int bufferSize)
 
         return new FSDataInputStream(
                 new BufferedFSInputStream(
-                        new CachingInputStream(inputStream, this, path, this.getConf(), statsMBean,
-                                               clusterManager.getSplitSize(), clusterManager.getClusterType()),
-                                                    CacheConfig.getBlockSize(getConf())));
+                        new CachingInputStream(inputStream, this, path,
+                                this.getConf(), statsMBean, clusterManager.getSplitSize(),
```
Comment (stagraqubole) [inline rubix-core/src/main/java/com/qubole/rubix/core/CachingFileSystem.java]: avoid doing changes which are not relevant to the PR, please revert all changes which are not relevant.

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java (hunk 2 above)
Comment (sakshia125) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: in retryConnection, there is the logic of closing and opening the bookKeeperClient.transport too.

Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: thats fine but why do we need to do bookKeeperClient.getCacheStatus once before calling retryConnection? We should be fine just calling retryConnection

Comment (sakshia125) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: no, because this in sort of a wraparound around bookKeeperClient. Calls to getCacheStatus will go through this. So why do we need to retry everytime there's a call to getCacheStatus

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        }
+    }
+
+    //Assuming that transport from bookKeeperClient is already open
+    private <V> V retryConnection(Callable<V> callable)
+            throws TException
+    {
+        int errors = 0;
+        try {
+            return callable.call();
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: again code duplication ... write it such that the while loop below handles it all

File: rubix-core/src/main/java/com/qubole/rubix/core/RemoteReadRequestChain.java [hunk 6]
```diff
@@ -29,10 +29,11 @@
 
 /**
  * Created by stagra on 4/1/16.
- *
+ * <p>
```
Comment (stagraqubole) [inline rubix-core/src/main/java/com/qubole/rubix/core/RemoteReadRequestChain.java]: whats this change? please revert

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                @Override
+                public List<Location> call()
+                        throws TException
+                {
+                    return bookKeeperClient.getCacheStatus(remotePath, fileLength, lastModified, startBlock, endBlock, clusterType);
+                }
+            });
+        }
+        catch (TException e) {
+            LOG.info("Could not get cache status from server");
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: I dont think we need the try-catch here, this method can throw TException

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                public Void call()
+                        throws Exception
+                {
+                    bookKeeperClient.setAllCached(remotePath, fileLength, lastModified, startBlock, endBlock);
+                    return null;
+                }
+            });
+        }
+        catch (Exception e1) {
+            LOG.info("Could not update BookKeeper about newly cached blocks: " + Throwables.getStackTraceAsString(e1));
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: Same here, no try-catch, let the exception bubble up and let inputStream handle it

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        }
+        catch (Exception e) {
+            LOG.info("Retrying connection" + e.getStackTrace().toString());
+        }
+
+        bookKeeperClient.transport.close();
+
+        while (errors < maxRetries) {
+            try {
+                bookKeeperClient.transport.open();
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: can you also verify that open() on transport that had been closed is fine? Maybe write a simple java program to do this

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperClient.java [hunk 10]
```diff
@@ -53,14 +55,16 @@ public void close()
...
     {
         TTransport transport;
-        transport = new TSocket("localhost", CacheConfig.getServerPort(conf));
+        transport = new TSocket("localhost", CacheConfig.getServerPort(conf), CacheConfig.getClientTimeout(conf));
+
         transport.open();
 
         BookKeeperClient client = new BookKeeperClient(transport);
-        return client;
+        RetryingBookkeeperClient retryingBookkeeperClient = new RetryingBookkeeperClient(client, CacheConfig.getMaxRetries(conf));
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperClient.java]: Forgot to catch this before, but creating the BookKeeperClient seems redundant. We can very well move the logic of BookKeeperClient.getCacheStatus etc into RetryinBookKeeperClient.

Do we need to keep BookKeeperClient around for any other use? If not lets get rid of it and only keep RetryinBookKeeperClient.

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        extends BookKeeperService.Client
+        implements Closeable
+{
+    private static final Logger LOG = LoggerFactory.getLogger(RetryingBookkeeperClient.class);
+    private int maxRetries;
+    private BookKeeperClient bookKeeperClient = null;
+
+    /**
+     * List of causes for TTransportException
+     * ALREADY_OPEN
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: Why are these here?

File: rubix-core/src/main/java/com/qubole/rubix/core/CachingInputStream.java [hunk 12]
```diff
@@ -18,10 +18,9 @@
 import com.google.common.util.concurrent.ListenableFuture;
 import com.google.common.util.concurrent.ListeningExecutorService;
 import com.google.common.util.concurrent.MoreExecutors;
-import com.qubole.rubix.bookkeeper.BookKeeperClient;
 import com.qubole.rubix.bookkeeper.Location;
+import com.qubole.rubix.bookkeeper.RetryingBookkeeperClient;
```
Comment (stagraqubole) [inline rubix-core/src/main/java/com/qubole/rubix/core/CachingInputStream.java]: Looking at all the issues we have seen regarding dependency orders, we should use interfaces instead.
So instead of using RetryingBookkeeperClient in this class we should be using BookKeeperService.Client.

That will help us in future too if we extract out thrift interfaces to spi.

File: rubix-core/src/main/java/com/qubole/rubix/core/RemoteReadRequestChain.java [hunk 13]
```diff
@@ -13,7 +13,7 @@
 package com.qubole.rubix.core;
 
 import com.google.common.base.Throwables;
-import com.qubole.rubix.bookkeeper.BookKeeperClient;
+import com.qubole.rubix.bookkeeper.RetryingBookkeeperClient;
```
Comment (stagraqubole) [inline rubix-core/src/main/java/com/qubole/rubix/core/RemoteReadRequestChain.java]: Same like above, better to change to BookKeeperService.Client

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+        int errors = 0;
+
+        while (errors < maxRetries) {
+            try {
+                if (!transport.isOpen()) {
+                    try {
+                        transport.open();
+                    }
+                    catch (Exception e1) {
+                        LOG.info("Error while reconnecting");
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: this could end up in infinite loop, let the transport.open's exception get caught outside so that errors are incremented and we can break

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 15]
```diff
@@ -97,21 +97,17 @@ public Void call()
...
+                    transport.open();
                 }
                 return callable.call();
             }
             catch (Exception e) {
                 LOG.info("Error while connecting" + e.getStackTrace().toString());
                 errors++;
             }
-            transport.close();
+            if (transport.isOpen()) {
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: why close here? we will close on when close() is called on this object

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java [hunk 16]
```diff
@@ -97,21 +97,17 @@ public Void call()
...
             try {
                 if (!transport.isOpen()) {
-                    try {
-                        transport.open();
-                    }
-                    catch (Exception e1) {
-                        LOG.info("Error while reconnecting");
-                        continue;
-                    }
+                    transport.open();
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java]: also, isnt this unncessary? The assumption says transport is already open and also when client is created, transport is opened explicitly as per assumption
